src/DotNetWorker.Grpc/FunctionMetadata/DefaultFunctionMetadataProvider.cs

```diff
@@ -0,0 +1,121 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading.Tasks;
+using System.Text.Json;
+using Microsoft.Azure.Functions.Worker.Grpc.Messages;
+
+namespace Microsoft.Azure.Functions.Worker
+{ 
+    internal class DefaultFunctionMetadataProvider : IFunctionMetadataProvider
+    {
+        private const string FileName = "functions.metadata";
+        private JsonSerializerOptions deserializationOptions;
+
+        public DefaultFunctionMetadataProvider()
+        {
+            deserializationOptions = new JsonSerializerOptions();
+            deserializationOptions.PropertyNameCaseInsensitive = true;
+        }
+
+        public virtual async Task<ImmutableArray<RpcFunctionMetadata>> GetFunctionMetadataAsync(string directory)
+        {
+            string metadataFile = Path.Combine(directory, FileName);
+
+            if (File.Exists(metadataFile))
+            {
+                using (var fs = File.OpenRead(metadataFile))
+                {
+                    // deserialize as json element to preserve raw bindings
+                    var jsonMetadataList = await JsonSerializer.DeserializeAsync<JsonElement>(fs);
+
+                    var functionMetadataResults= new List<RpcFunctionMetadata>(jsonMetadataList.GetArrayLength());
+
+                    foreach (var jsonMetadata in jsonMetadataList.EnumerateArray())
+                    {
+                        var functionMetadata = JsonSerializer.Deserialize<RpcFunctionMetadata>(jsonMetadata.GetRawText(), deserializationOptions);
+
+                        if (functionMetadata is null)
+                        {
+                            throw new NullReferenceException("Function metadata could not be found.");
+                        }
+
+                        // hard-coded values that are checked for when the host validates functions
+                        functionMetadata.IsProxy = false;
+                        functionMetadata.Language = "dotnet-isolated";
+                        functionMetadata.FunctionId = Guid.NewGuid().ToString();
+
+                        var rawBindings = GetRawBindings(jsonMetadata);
+
+                        foreach (var binding in rawBindings.EnumerateArray())
+                        {
+                            functionMetadata.RawBindings.Add(binding.GetRawText());
+
+                            BindingInfo bindingInfo = CreateBindingInfo(binding);
+
+                            binding.TryGetProperty("name", out JsonElement jsonName);
+
+                            functionMetadata.Bindings.Add(jsonName.ToString(), bindingInfo);
+                        }
+
+                        functionMetadataResults.Add(functionMetadata);
+                    }
+
+                    return functionMetadataResults.ToImmutableArray();
+                }
+            }
+
+            return ImmutableArray<RpcFunctionMetadata>.Empty;
```

Comment:
Should [we throw](https://github.com/Azure/azure-functions-dotnet-worker/pull/825) when the file is missing?

Reply:
good call